src/program.cpp

```diff
@@ -47,14 +49,292 @@ struct membuf : public std::streambuf {
+                auto kernel = helper->strings[inst->words[5]];
+                auto ordinal = helper->constants[inst->words[6]];
+                auto spec_id = helper->constants[inst->words[7]];
+                auto size = helper->constants[inst->words[8]];
+                std::string arg_name;
+                if (inst->num_operands == 9) {
+                    arg_name = helper->strings[inst->words[9]];
+                }
+                auto kind = inst_to_arg_kind(ext_inst);
+                kernel_argument arg = {arg_name, ordinal, 0,       0,   0,
+                                       0,        kind,    spec_id, size};
+                helper->binary->add_kernel_argument(kernel, std::move(arg));
+                break;
+            }
+            case NonSemanticClspvReflectionSpecConstantWorkgroupSize: {
+                // Reflection encodes all three spec ids in a single
+                // instruction.
+                auto x_id = helper->constants[inst->words[5]];
+                auto y_id = helper->constants[inst->words[6]];
+                auto z_id = helper->constants[inst->words[7]];
+                helper->binary->add_spec_constant(
+                    spec_constant::workgroup_size_x, x_id);
+                helper->binary->add_spec_constant(
+                    spec_constant::workgroup_size_y, y_id);
+                helper->binary->add_spec_constant(
+                    spec_constant::workgroup_size_z, z_id);
+                break;
+            }
+            case NonSemanticClspvReflectionSpecConstantGlobalOffset: {
+                // Reflection encodes all three spec ids in a single
+                // instruction.
+                auto x_id = helper->constants[inst->words[5]];
+                auto y_id = helper->constants[inst->words[6]];
+                auto z_id = helper->constants[inst->words[7]];
+                helper->binary->add_spec_constant(
+                    spec_constant::global_offset_x, x_id);
+                helper->binary->add_spec_constant(
+                    spec_constant::global_offset_y, y_id);
+                helper->binary->add_spec_constant(
+                    spec_constant::global_offset_z, z_id);
+                break;
+            }
+            case NonSemanticClspvReflectionSpecConstantWorkDim: {
+                auto dim_id = helper->constants[inst->words[5]];
+                helper->binary->add_spec_constant(spec_constant::work_dim,
+                                                  dim_id);
+                break;
+            }
+            case NonSemanticClspvReflectionPushConstantGlobalOffset:
+            case NonSemanticClspvReflectionPushConstantEnqueuedLocalSize:
+            case NonSemanticClspvReflectionPushConstantGlobalSize:
+            case NonSemanticClspvReflectionPushConstantRegionOffset:
+            case NonSemanticClspvReflectionPushConstantNumWorkgroups:
+            case NonSemanticClspvReflectionPushConstantRegionGroupOffset: {
```

Comment:
I'm wondering whether a single `NonSemanticClspvReflectionPushConstant` instruction with an enum argument to specify the push constant wouldn't have been better. All code that needs to handle push constants generically will have to maintain an ever-growing list of instructions with the current scheme.

Reply:
I considered that when designing the instruction set, but it only really simplifies this switch statement. You still need the conversion switch statement in the other method. I felt this is easier to validate than also validating specific enum values. That said, this could change in future versions of the non-semantic instruction set.

Reply:
True. Enum validation could be done with just a range check but I'm not overly worried about this (we're unlikely to add new push constants every week) and, as you say, it's easy enough to evolve it.